### ISSUE TITLE
default constructible requirement for result<T>::value_type?

### ISSUE THREAD

**Issue by andreihu** [mnmlstc/core#45]
Hi,

When i try to instantiate `core::result<T>` on types which are not default constructible, i run into a compiler error. As far as i see this requirement is not listed in the documentation. Looking at the source code, i find this:

```
  result (::std::error_condition const& ec) :
    valid { not ec }
  {
    if (*this) { ::new (::core::as_void(this->val)) value_type(); }
    else { ::new (::core::as_void(this->cnd)) error_type(ec); }
  }
```

So if someone uses an 'ok' error_condition in the result<T> the underlying storage will be initialized with a default constructed instance of T. I see the underlying intent that since there is no error thus there should be a value, but this is also adding the default constructible requirement for type T, which i see a bit restrictive. Is there something that could be done, to use non default constructible types?

Thanks 
 Andras

**Comment by slurps-mad-rips** [comment]
Hi, sorry for the delay in a response. I'm usually good about these but have been dealing with personal issues as of late that have prevented me from dealing with this. So the constructor's don't state this in the 1.1 documentation (which is out of date), however the `error_condition` [assignment operator](http://mnmlstc.github.io/core/optional.html#result:T:::assign-operator__ErrorConditionEnum) does state that assigning an error condition that would result in a default constructed value type, it will still destruct the current internal value then default construct a new one.

And yes, this is very restrictive. `result<T>` was written quite some time ago, long before I figured out some different techniques that could be applied, along with a better approach to a type that handles both `core::expected<T>` and `core::result<T>` use cases.

As for how to handle this, my only recommendation would be to write a quick `result<T>` like wrapper using `core::variant` and hope for the best? Probably not what you wanted to here. Sorry.

**Comment by andreihu** [comment]
Hi, thanks for the clarification, it was my conclusion more or less too.

I am really interested in any developments regarding this topic (variant/optional/result/expected), since i think that these are extremely useful building blocks for API's. Are you planning any improvements on these?

As of result<T> could you please explain one more thing? It's using an error_condition where i think an error_code would be more correct. E.g. the value() method - when it throws- will be plain wrong since it takes the numeric value of the error_condition, constructs an error code and then a system_error and throws that. Error condition values are mapping directly to error code values in the system and the generic error_category, but this is not a universal rule. In my opinion there is no valid conversion from error_condition to error_code.

Regards
 Andras

**Comment by slurps-mad-rips** [comment]
Unfortunately, I can't actually do any improvements on these, without breaking a lot of people's code. While I am introducing breaking changes in 2.0, it's more so I could change the ABI than change the API. I have changed parts of the latter, to ease development on C++17 features that have been added (e.g., the "rarely empty variant").  That said, `core::variant` is meant to be a drop in replacement for Boost.Variant, but with better compile times and lambda visitation. I believe my implementation was one of the first to do the lambda overload trick, but I can't prove that beyond having written the thing back in summer of 2013. If you are looking for an improvement on the Boost.Expected type, I recommend taking a peak at the [Boost.Outcome](https://github.com/ned14/boost.outcome) library. It has an approach to error reporting that I quite like, though I think it does *too* much for the general case, and one could write an `outcome<T, E>` (to differentiate from the semantics provided by Boost.Expected), and specialize actions on certain error types. e.g., one could recreate the `expected<T>::raise` function by doing something like:

```cpp
template <class T>
[[noreturn]] void raise (outcome<T, std::exception_ptr> const& o) noexcept(false) {
    rethrow_exception(o.error()); // where error throws if the outcome is not invalid
}

template <class T>
[[noreturn]] void raise (outcome<T, std::error_code> const& o) noexcept(false) {
    throw std::system_error(o.error()); // where error throws if the outcome is not invalid
}

template <class E, class T>
auto expect (outcome<T, std::exception_ptr> const& o) noexcept(false) {
  try { raise(o); }
  catch (E const& e) { return e; }
}
```

A type like this could be *valid*, *invalid*, or *neither*, a third empty state that can permit the strong exception guarantee (this state is what is permitting the proposed `std::variant` to keep the *rarely empty* guarantee, which is an improvement over every other variant out there)

I realized that it was a mistake to make `core::result<T>` use `std::error_condition` about a year after I made it. Not only have I almost never used an `error_condition` outside of interacting with `result<T>`, but I like that `error_code` is fewer characters and platform specific. I've never had to translate error codes between platforms, because their use is usually wrapping C APIs on specific platforms. So obviously you *would* want a specific error code and category in those cases.

Unfortunately, there are no APIs that depend on the incorrect behavior of `result<T>`, and I am stuck either not changing it and letting MNMLTC Core die a slow death as better libraries are written by much smart people (which is actually it's original goal :sweat_smile:), or changing it as though Core has this last gasp of breath before it enters permanent maintenance mode.

When I wrote Core it was because I wanted C++14 proposals in C++11 *then*, and I was also working in the games industry. Even though our team moved to C++11, the plan was to stay there for a LONG TIME and never upgrade if possible (coincidentally, that company liquidated back in June of 2015). I always knew that at some point the usefulness of Core would be eclipsed by what is actually available from major implementations, but the idea was that it would provide an easier upgrade path for those few people who would be stuck on C++11. That said, I've seen a push by compiler vendors to put better features and tools available, so maybe the "not upgrading for many years" isn't as big of a problem for going forward. But I've never been one to predict the future of the language.

Anyhow, I could probably elaborate on all this more in a blog post and then post it to r/cpp, but that would probably be after I posted several other drafts I have in the pipeline, and I've been otherwise preoccupied.
